File: ai25btech11007/MatGeo/MatGeo_3/Codes/unit_vector.c
#include <stdio.h>

// Custom sqrt function using Newton-Raphson method
double mySqrt(double n) {
    double x = n;
    double root;
    int i;

    if (n == 0) return 0;

    // Iterate to improve accuracy
    for (i = 0; i < 20; i++) {
        root = 0.5 * (x + (n / x));
        x = root;
    }
    return root;
}

int main() {
    // Given vectors a = (1, -7, 7), b = (3, -2, 2)
    double a[3] = {1, -7, 7};
    double b[3] = {3, -2, 2};
    double n[3];   // cross product
    double magnitude, unit[3];

    // Cross product: n = a × b
    n[0] = a[1]*b[2] - a[2]*b[1]; // i component
    n[1] = a[2]*b[0] - a[0]*b[2]; // j component
    n[2] = a[0]*b[1] - a[1]*b[0]; // k component

    // Magnitude of vector n using custom sqrt
    magnitude = mySqrt(n[0]*n[0] + n[1]*n[1] + n[2]*n[2]);

    // Unit vector = n / |n|
    unit[0] = n[0] / magnitude;
    unit[1] = n[1] / magnitude;
    unit[2] = n[2] / magnitude;

    // Output results
    printf("Vector perpendicular to both a and b: (%.2f, %.2f, %.2f)\n", n[0], n[1], n[2]);
    printf("Unit vector: (%.6f, %.6f, %.6f)\n", unit[0], unit[1], unit[2]);

    return 0;
}

File: ai25btech11007/MatGeo/MatGeo_3/Codes/unit_vector.py
import numpy as np
import matplotlib.pyplot as plt
import os

# Create figs directory if not exists
os.makedirs("figs", exist_ok=True)

# Define vectors
a = np.array([1, -7, 7])
b = np.array([3, -2, 2])

# Cross product
n = np.cross(a, b)

# Magnitude of cross product
magnitude = np.linalg.norm(n)

# Unit vector
unit = n / magnitude

# Print results
print("Vector a:", a)
print("Vector b:", b)
print("Cross product (a × b):", n)
print("Magnitude of cross product:", magnitude)
print("Unit vector perpendicular to a and b:", unit)

# Save results to text file
results_path = "figs/results.txt"
with open(results_path, "w") as f:
    f.write("Vector a: " + str(a) + "\n")
    f.write("Vector b: " + str(b) + "\n")
    f.write("Cross product (a × b): " + str(n) + "\n")
    f.write("Magnitude of cross product: " + str(magnitude) + "\n")
    f.write("Unit vector perpendicular to a and b: " + str(unit) + "\n")

print(f"Results saved at: {os.path.abspath(results_path)}")

# Plotting
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')

# Plot vectors a, b, and unit vector
ax.quiver(0, 0, 0, a[0], a[1], a[2], color='r', label='a')
ax.quiver(0, 0, 0, b[0], b[1], b[2], color='g', label='b')
ax.quiver(0, 0, 0, unit[0], unit[1], unit[2], color='b', label='Unit vector')

# Set axes labels
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_zlabel('Z')
ax.legend()
ax.set_title("Vectors a, b and perpendicular unit vector")

# Save figure
save_path = "figs/vectors.png"
plt.savefig(save_path)
print(f"Figure saved at: {os.path.abspath(save_path)}")

plt.show()
